### ISSUE TITLE
Add HTTP/2 support to the HTTP server

### ISSUE THREAD

**Issue by nyh** [scylladb/seastar#734]
HTTP/2 added a number of interesting features beyond HTTP 1.1, such as request multiplexing and header compression. It would be good if Seastar's builtin HTTP server implementation, which is currently limited to HTTP 1.1, also implemented HTTP/2. We probably want this support for both plain-text and encrypted (TLS) connections.

Seastar-using projects are already waiting for this - e.g., https://github.com/scylladb/scylla/issues/6113.

**Comment by jewertow** [comment]
Hello @nyh, could I work on this issue?

**Comment by nyh** [comment]
I don't think anyone else is working on this issue at the moment, so the answer is probably yes.
@wmitros has been working recently on improving Seastar's HTTP 1.1 support (fixing various parsing bugs, supporting chunked-encoding, etc.) but I don't know if he was planning to "expand" to HTTP 2 in the near future.

**Comment by wmitros** [comment]
I was thinking about expanding to HTTP 2 at some unspecified time in the future, but I don't think there's anything that should be preventing you from taking it up right now @jewertow

**Comment by jewertow** [comment]
Thank you for the answer. I need few weeks to investigate and implement this feature.

**Comment by dimaqq** [comment]
I wonder if Jacek is still working on this or if he gave up 😅

**Comment by dimaqq** [comment]
Poke

**Comment by jewertow** [comment]
No, I didn't work on that, because the project for which it was needed, used the proxygen library to handle http/2 and didn't use seastar at all. I'm sorry, I shouldn't have declared that I will work on it without being sure.

**Comment by mykaul** [comment]
I'd skip and go directly to HTTP/3.

**Comment by nyh** [comment]
> I'd skip and go directly to HTTP/3.

I agree that implementing https://datatracker.ietf.org/doc/html/rfc9000 and https://datatracker.ietf.org/doc/html/rfc9114 will be cool.

Just note that HTTP/3 is much newer and support for it in various languages (including my favorite playground, Python) is much less mature than support for HTTP/2, so it may be harder to write clients to the HTTP/3 server. But I'm sure that the client availability will improve in the next few years.

**Comment by muthu90tech** [comment]
Hi @mykaul / @nyh , I would like to work on the HTTP/3 support in that case, would like to add incremental changes
I can repurpose the http2 frame parsing to http3 frame parsing: https://github.com/scylladb/seastar/pull/2886
do you recommend I work and commit to a dedicated branch in your repo? thank you

I noticed HTTP clients for dynamoDB still does not support HTTP/3, unsure how long that would take, considering that, if that is alright I can continue to push changes for HTTP2 in a branch , please let me know thx.

**Comment by muthu90tech** [comment]
Ping for this again @nyh / @mykaul I am continuing to work on the HTTP2 support locally. 
My plan for this was to 
1. work in a dedicated branch in seastar repo.
2. Pull request smaller iteration into the branch get feedback
3. Test and merge into the main after performance tests

For eg: https://github.com/scylladb/seastar/pull/2886

**Comment by mykaul** [comment]
I'm not sure why you are pinging us - feel free to work on this. The usual way to do it is work on your branch and then send a pull request, NOT to commit in this repo.

**Comment by nyh** [comment]
@muthu90tech when I created this issue, HTTP/3 did not yet exist, so I did not need to make the choice between HTTP/2 and HTTP/3. Now, you do, @mykaul and I gave a few pros and cons for each above, but neither of us have any concrete plans to use either immediately - so we don't really insist on either and it's your choice what you want to implement - what is more interesting for you, or sounds more useful to you, or perhaps even - easier to do.

I think that you'll need to work in your own branch until you're ready sent a pull request with the completed code.

Ideally, if several people are working on the same project, they would be reviewing each other's work in smaller iterations, so somebody doesn't spend a month pursuing a direction that the would-be-reviewer hates. However, in this case I don't think that I or @mykaul or anyone else has the "bandwidth" to actually do many review iterations over your code, nor do we have any immediate plans to use this feature and waiting anxiously to test it. So if you don't find any other volunteer to review your work incrementally (I see above several people who expressed interest in this issue and might be interested to review your work) - then I think you'll have no choice but to work on your patches yourself, and only when you consider the result to be correct, reasonably performant and tested, and you fixed your own bugs and design mistakes - then send the code for review.

Of course, if before then you have questions - or want someone to look at specific not-yet-complete code and answer something about it, we can definitely do that as well.

**Comment by muthu90tech** [comment]
Sounds good @nyh , I will work on my branch (http2). Thank you
